x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/show/ShowFunctions.java

```diff
@@ -20,24 +23,64 @@
 import org.elasticsearch.xpack.ql.tree.NodeInfo;
 import org.elasticsearch.xpack.ql.tree.Source;
 
+import java.lang.reflect.Constructor;
 import java.util.ArrayList;
 import java.util.Comparator;
 import java.util.List;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 
+import static org.elasticsearch.xpack.ql.type.DataTypes.BOOLEAN;
 import static org.elasticsearch.xpack.ql.type.DataTypes.KEYWORD;
 
 public class ShowFunctions extends LeafPlan {
 
     private final List<Attribute> attributes;
 
+    public record ArgSignature(String name, String type, String description, boolean optional) {}
+
+    public record FunctionSignature(String name, List<ArgSignature> args, String returnType, String description, boolean variadic) {
```

Comment:
Better to 'promote' this to `FunctionDescription` and move that as part of the `EsqlFunctionRegistry` so that it gets created each time a function definition gets created.
Move the reflection/annotation logic there along with some fallbacks in case the annotations are missing (which we should catch through some basic tests that check the functions registered and that they all have the proper description/annotation).

ShowFunction itself should just call the functionregistry and iterate over the data.

Reply:
Good point, this logic belongs to `EsqlFunctionRegistry`, I'll move it there for now.
Moving it into `FunctionDescription` and generating it only once is a bit more complicated, because that class is part of QL. It's probably a good candidate for a follow-up PR, to keep this one small.

Reply:
> however please time box it

Definitely!

Reply:
I think you meant FunctionDefinition, FunctionDescription would be a new class specific to ESQL.

Reply:
Yes, sorry, `FunctionDefinition`